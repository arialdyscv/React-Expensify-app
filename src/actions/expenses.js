import uuid from "uuid";
import { App, DB } from "../firebase/firebase"; 
import { 
  getDatabase, ref, get, set, update, remove, push, onValue, off, child, 
  onChildRemoved, onChildChanged, onChildAdded } from 'firebase/database';

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense
});

// : {
//   id: uuid(),
//   description,
//   note,
//   amount,
//   createdAt,
// },

//ADDS to the Database
export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = 0,
    } = expenseData;

    const expense = { description, note, amount, createdAt};

    return push(ref(DB,'expenses'), expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });

  };
};

// REMOVE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

//REMOVES from Database
export const startRemoveExpense = ( { id } = {}) => {
  return (dispatch) => {

    return remove(ref(DB), `expenses/${id}`).then(() => {
      dispatch(removeExpense({ id }));

    });
  }
};

// EDIT
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});

export const startEditExpense = ( id, updates ) => {
  return (dispatch) => {

    return update(ref(DB, `expenses/${id}`), updates).then(() => {
      dispatch(editExpense(id, updates));
    });
  }
}

//SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: "SET_EXPENSES",
  expenses
});

//GETS list of expenses from Database
export const statSetExpenses = () => {
  return (dispatch) => {
    const dbRef = ref(DB);
    return get(child(dbRef, 'expenses')).then((snapshot) => {
      if(snapshot.exists()) {
        const expenses = []
        snapshot.forEach((childSnap) => {
          expenses.push({
            id: childSnap.key,
            ...childSnap.val(),
          });
        });
        dispatch(setExpenses(expenses))
      } else {
        console.log('No Expenses')
      }
    }).catch((error) => {
      console.log(error);
    });
  }
}
