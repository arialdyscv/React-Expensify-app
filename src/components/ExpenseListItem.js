import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ id, description, amount, createdAt, note } ) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>
            {numeral(amount / 100).format('$0,0.00')} 
            - 
            {moment(createdAt).format('Do / MMM / YY')}
        </p>
        {note ? <p>{note}</p> : ''}
        
    </div>
);


export default ExpenseListItem;

