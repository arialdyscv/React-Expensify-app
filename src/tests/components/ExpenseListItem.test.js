import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import ExpenseListItem from '../../components/ExpenseListItem';


test('should render ExpenseListItem with expense index 0', () =>{
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
    expect(wrapper).toMatchSnapshot()
});


