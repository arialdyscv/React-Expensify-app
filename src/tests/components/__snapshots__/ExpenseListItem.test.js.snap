// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with expense index 0 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      water bill
    </h3>
  </Link>
  <p>
    $15.00
    -
    31st / Dec / 69
  </p>
</div>
`;
